pkg/agent/billing/send.go

```diff
@@ -0,0 +1,147 @@
+package billing
+
+// Logic responsible for sending billing events by repeatedly pulling from the eventQueue
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/neondatabase/autoscaling/pkg/billing"
+	"github.com/neondatabase/autoscaling/pkg/util"
+)
+
+type eventSender struct {
+	client            billing.Client
+	config            *Config
+	metrics           PromMetrics
+	queue             eventQueuePuller[billing.IncrementalEvent]
+	collectorFinished util.CondChannelReceiver
+
+	// lastSendDuration tracks the "real" last full duration of (eventSender).sendAllCurrentEvents().
+	//
+	// It's separate from metrics.lastSendDuration because (a) we'd like to include the duration of
+	// ongoing calls to sendAllCurrentEvents, but (b) we don't want the bias towards lower durations
+	// that comes with that.
+	//
+	// Here's some more detail:
+	//
+	// To make sure that long-running sendAllCurrentEvents() loops show up in the metrics while
+	// they're still running, we want to periodically update metrics.lastSendDuration before the
+	// loop has finished. A side-effect of doing this naively is that the gauge will sometimes
+	// return durations that are much shorter than the *actual* previous send loop duration.
+	//
+	// In order to fix this, we store that *actual* previous duration in this field, but and only
+	// update the metric when either (a) the loop is done, or (b) the duration so far is already
+	// longer than the previous one.
+	//
+	// This means that we remove the bias towards shorter durations, at the expense of sometimes
+	// returning higher durations for too long. IMO that's ok, and we'd rather have our metrics give
+	// a pessimistic but more accurate view.
+	lastSendDuration time.Duration
+}
+
+func (s eventSender) senderLoop(logger *zap.Logger) {
```

Comment:
I would want this to have a context, and be able to handle shutdown in some clean way.

Reply:
So it _mostly_ does that with the `collectorFinished` signal — the reason it doesn't just take a context is because we could miss some trailing events (e.g. the "collect" thread receives the shutdown notification & packages off the last set of events, but if the "send" thread has already exited, we'll drop some)